juju/juju
# [JUJU-3333] Users removal triggers the removal of model access

After enabling user names to be reused (#15266 ) it has been observed that for those recreated users that where previously granted with access permissions to other models, these permissions are automatically recreated. This is a misbehaviour that is corrected in this PR by removing all the permissions a user is granted when the user is removed from the controller. This makes possible to have a clear list of permissions when a user is recreated avoiding this unexpected behaviour.

## Checklist

*If an item is not applicable, use `~strikethrough~`.*

~- [ ] Code style: imports ordered, good names, simple structure, etc~
~- [ ] Comments saying why design decisions were made~
~- [ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
~- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Try to replicate an scenario for a recreated user who was granted some permissions.

```sh
juju add-user mrx
User "mrx" added
Please send this command to mrx:
    juju register MHsTA21yeDBFDC5bZmQ0MjpjZmExOmFkZjc6NjI3NDoyMTY6M2VmZjpmZTE1OjZiYzldOjE3MDcwExMxMC4yMjAuMTguMjIyOjE3MDcwBCC9TIZM91FQZPJ2bjnICsB4rcyUlcpmPXrxgpQpxJx5NRMJbG9jYWx0ZXN0EwAA

"mrx" has not been granted access to any models. You can use "juju grant" to grant access.
```
Grant user access
```sh
juju add-model test
juju grant mrx admin test
juju users test
Name    Display name  Access  Last connection
admin*  admin         admin   36 minutes ago
mrx                   admin   never connected
```
Destroy the user and recreate
```sh
juju remove-user mrx --yes
users test
Name    Display name  Access  Last connection
admin*  admin         admin   37 minutes ago

juju add-user mrx
User "mrx" added
Please send this command to mrx:
    juju register MHsTA21yeDBFDC5bZmQ0MjpjZmExOmFkZjc6NjI3NDoyMTY6M2VmZjpmZTE1OjZiYzldOjE3MDcwExMxMC4yMjAuMTguMjIyOjE3MDcwBCC9TIZM91FQZPJ2bjnICsB4rcyUlcpmPXrxgpQpxJx5NRMJbG9jYWx0ZXN0EwAA

"mrx" has not been granted access to any models. You can use "juju grant" to grant access.
```

Now if you check the users for model `test`, no permissions are set
```sh
users test
Name    Display name  Access  Last connection
admin*  admin         admin   37 minutes ago
```
